### ISSUE TITLE
Have a schema for bindings, and report error on miss-typed bindings.

### ISSUE THREAD

**Issue by mhevery** [angular/angular#2014]
Quick, can you spot an error? (Currently it fails silently.)

```html
<ul>
  <li *ng-for="item in items">{{item}}</li>
</ul>
```
There are two issues:
- It is `for of` not `for in`
- It needs `var item` or `#item` instead of `item`

The current binding expands to 

```html
<ul>
  <template [ng-for]="item" [ng-for-in]="items">
    <li>{{item}}</li>
  </template>
</ul>
```

but it should have been:

```html
<ul>
  <template ng-for #item [ng-for-of]="items">
    <li>{{item}}</li>
  </template>
</ul>
```

- [x] it fails to match `[ng-for][ng-for-of]` and turns the binding to element property bindings (ie monkey patches the properties on template element) (This should have failed because it should have noticed that the element does not have `ngFor` or `ngForOf` and hence should have prevent the monkey patch)
- [ ] it then fails to notice that `item` is not defined (because we are trying to define it) (This should have failed!)

# GOAL
- Identify typos in the HTML bindings and throw errors to hint to developers.

# Strategy

The way to do this properly is to have a schema for which properties are allowed on which DOM elements. Currently we can lazy compute such as schema for DOM elements. But we can not compute them for WebComponents. So the proposal is to WebComponents publish their own schema.

```html
<div [title]="exp"></div>
<svg>
  <g [x]="exp">
<svg>
```

We can lazy compute that `title` is a property of `div` (by noticing that `title` is a property of the `div` element). This can be done in Compiler during the compilation step. The result would be something like:

```
var schema = {
  'div': { 'title': 'title' },
  'svg:g': { 'x': '@x' }
}
```

The above would tell us that it is OK to bind to `title` to internal property name `title` on a `div`. (We need this as some properties don't properly rename see `{ 'inner-html': 'innerHTML', 'readonly': 'readOnly', 'tabindex': 'tabIndex'}` )

We could also tell that `g` is instance of `SVGElement` and in that case we bind to the `x` attribute rather then the default `x` properties. 

But this does not work with WebComponents because during the compilation the element is in `template` tag and is not active, as a result we can not query its properties.

```html
<google-youtube [video-id]="exp">
```

there is no way for Angular to know if `videoId` is a valid, as a result we need to configure this:

```javascript
@View({
  webComponents: {
    'google-youtube': {'!extends': 'div', 'video-id': 'videoId'}
  }
})
```
NOTE: `!extends` is used to say that all mappings from the `div` should be included.

By adding this schema definition Angular will be able to correctly resolve web-component assignments.

# Resolution Rules

```html
<div [foo]="exp" bar="literal">
```

1. `foo` (property binding)
  - Is in schema => then assign `foo` property to element
  - is in schema and is a directive => assign to `foo` property and to the directive
  - is not in schema but it has a directive => assign to directive only
  - is not in schema and does not have a directive => Throw an error
2. `bar` (just a literal)
  - Is in schema => no-op (we leave the property as is)
  - is in schema and is a directive => assign the literal one-time to the directive
  - is not in schema but it has a directive => assign the literal one-time to the directive
  - is not in schema and does not have a directive => Throw an error. (Implication is that if a component wants to use attributes for CSS which are not part of DOM, they will have to be declared ahead of time in the schema.)

# Error handling.

With the above rules the example will produce these errors:

```html
<ul>
  <li *ng-for="item in items">{{item}}</li>
</ul>
```

```html
<ul>
  <template [ng-for]="item" [ng-for-in]="items">
    <li>{{item}}</li>
  </template>
</ul>
```

* `Can not bind 'item' to 'ng-for' because 'ng-for' is not a property of [TemplateElement, NgFor].`
* `Can not bind `items` to 'ng-for-in' because 'ng-for' is not property of [TemplateElement, NgFor].`

**Comment by ewinslow** [comment]
Thanks for working on this, Misko! This is really the first developer-happiness-related issue I noticed when I tried out the dev preview, so glad to see it is being addressed and thought about. I like the idea of being able to declare schemas. Do we also need to be able to declare exposed events for the same reasons given?

There are a few more magic symbols introduced here which makes me a bit nervous. Can we try to avoid adding more symbols to the API? I'm speaking here of `@` for "bind to attribute" and `!extends`. Symbols are asking to become "Angular WAT" material for sure...

I'd also love to encourage more user-centered error messaging, if at all possible. The current error messages are better than nothing, but otherwise meaningless to me as a humble developer (you might also have some typos? If so, that's not helping either). What I'm looking for is something more straightforward like:

> You wrote `*ng-for="item in items"`, did you mean `*ng-for="#item of items"`?

I realize that guessing developer intent is a bit hard to do in the general case, and I apologize for not having a ready suggestion to that end, but I suspect forgetting to use `#` or `var` to declare variables is going to be a rather common issue, so it'd be great to get the most helpful messages we can protecting developers against that pitfall.

**Comment by vicb** [comment]
@naomiblack Have you unassigned me by error or is someone else working on this ?

**Comment by mhevery** [comment]
@pkozlowski-opensource Can you also think about Event schema as well?

**Comment by pkozlowski-opensource** [comment]
@mhevery so I've started to dig into the topic and here are some thoughts:
* I agree with @ewinslow that `@` is a bit "magical" - on top of this we've got already a system in place where we can specify that sth is an attribute binding, so my proposal would be `'svg:g': { 'x': 'attr.x' }`
* I'm not sure how I feel about sticking `'!extends'` into the properties map - maybe it deserves its own key?
* when it comes to extending - maybe we should be able to specify both an element name and / or its class (to be tested with `instanceof`) as probably many elements will have the same properties. 
* since we are doing properties re-mapping we will be able to remove the "magical" mapping ('foo: bar') from the `properties`, no?
* When it comes to events - I guess we can't find them out during runtime, so we would have to "hard-code" a list of all the events?

With my current understanding the (slightly) modified proposal would be:

```javascript
@View({
  schema: { //not sure about the "schema" name here...
    'google-youtube': {
        extends: ['div'], //I guess we can allow extending multiple elements, no?
        properties: {
                'video-id': 'videoId'  //this can get quite verbose.... same problem as properties in @Directive
        },
        events: ['play', 'stop', ...]     
    }
  }
})
```` 

Could we discuss those items at some point?

**Comment by pkozlowski-opensource** [comment]
Ah, regarding the implementation - I'm leaning towards adding a new, dedicated compilation step (one that would execute after all the directives and properties are parsed) or sticking this logic into `ProtoViewBuilder`. Any thoughts on the best place for the schema-verification logic and / or other implementation hints?

**Comment by pkozlowski-opensource** [comment]
OK, so did good progress on this today, have the first implementation workish (at least in TS/JS) but additional checks uncovered 2 interesting corner cases:

1) Given

```java
@View({
	template: '<div [foo]></div>'
	directives: []
}) 
```

shell we throw? I would say YES (no `foo` property on an element, no directives)

2) Given 

```java
@Directive({
	selector: '[foo]'
})
class Foo {	}

@View({
	template: '<div [foo]></div>'
	directives: [Foo]
})
```

I would say YES (no `foo` property on an element, nor any matching directives)

If we do those changes, the syntax like `<template [ng-for] #item [ng-for-of]="items">` will throw - people will have to write `<template ng-for #item [ng-for-of]="items">`

@mhevery could I have your opinion on those 2 cases?

**Comment by pkozlowski-opensource** [comment]
The same pb for the existing switch directive. Today this works: `<template [ng-switch-default]><li>when default;</li></template>` but if we want to be more strict, it would have to be `<template ng-switch-default><li>when default;</li></template>`

**Comment by mhevery** [comment]
1) is a clear YES.

2) the `foo` attribute exists and there is a `[foo]` selector which matches it. Your concern is that attribute `[foo]` implies expression where there is none. Which I think is an error.

**Comment by kvantetore** [comment]
Considering #2600, I guess the schema for events would also have to be an map rather than a list so we can override the default camel casing.

**Comment by pkozlowski-opensource** [comment]
@kvantetore right, to be considered.

**Comment by yjbanov** [comment]
Would it be reasonable to expect that this will be closed in alpha-29 or should we bump it to alpha-30?

**Comment by pkozlowski-opensource** [comment]
@yjbanov I will make some progress in .29 but probably complete all items only in alpha.30, so it is reasonable to move this issue.

**Comment by pkozlowski-opensource** [comment]
@mhevery I'm progressing on the schema support implementation but while coding / thinking some more there are 2 items that need discussing, IMO:
* "global" vs. per `@View` schema - last time we've discussed the concern was that if we declare is as "global" (application-wide) we would have problems with regular directives (since those can be specified per-view). But we [can distinguish between binding to element and directive properties](https://github.com/angular/angular/blob/3e6503789ff445b765996bcb530e4ad88ace19e0/modules/angular2/src/render/dom/view/proto_view_builder.ts#L363) and we know how to query directive properties at run-time. So we only need schema for HTML elements (regular and custom) really and those are global
* Interestingly, it seems like we can [detect web components at runtime](http://stackoverflow.com/questions/27334365/how-to-get-list-of-registered-custom-elements) so it seems like we could query custom properties as well (!). I need to experiment with this a bit more but it seems like we could call `document.createElement('my-el')` during compilation and see if we are dealing with a custom element + eventually query its properties. In this case we would need schema for events / custom attributes only.

**Comment by mhevery** [comment]
Sounds reasonable. Let's give that a try first and then revisit if it will not be sufficient.

**Comment by naomiblack** [comment]
Raising prioirity -- this feature is needed by internal users to remove the hack workaround in #3019. @pkozlowski-opensource @mhevery can you sync on this today and update with status?

**Comment by pkozlowski-opensource** [comment]
@naomiblack sorry, just saw your message right now. I'm back from the trainings and can resume work on this starting tomorrow.

**Comment by pkozlowski-opensource** [comment]
Status update: sent https://github.com/angular/angular/pull/3225 for the basic infrastructure impl. It needs proper review + discussion about next steps.

**Comment by pkozlowski-opensource** [comment]
I'm almost here, working on plugging the registry into the whole system. Not sure if we will be able to land  #3225 today. 

@mhevery @tbosch could we have a quick chat today about steps to make it fully work for Dart?

**Comment by pkozlowski-opensource** [comment]
@mhevery so, we've got already functionality for detecting missing bindings and the initial schema support landed via  d894aa9101d5662758c4ad4d66d8f1be9fc562a4.

I think that it is time to replace this big issue with smaller, more-focused ones:
* handling custom elements
* events
* attributes
* hand-written schema for Dart
* ...?

I can propose / open those issues and run them through you.

**Comment by pkozlowski-opensource** [comment]
@mhevery With basic schema support done in d894aa9 I've opened smaller and focused issues to add missing pieces of elements schema support:
* handling custom elements: #3424
* attributes #3425
* events: #3426
* hand-written schema for Dart: #3427

Closing this one in favor of the newly opened issues, would be great if you could triage those to prioritize work on next steps.
